### PR TITLE
fix def collector for impl trait

### LINKED ISSUE #77329: Stable ICE: assertion failed: `(left == right)`
### Code

```Rust
fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| { x } }]> {
    std::iter::empty()
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.46.0 (04488afe3 2020-08-24)
rustc 1.47.0-beta.5 (9f0e6fa94 2020-09-29)
rustc 1.48.0-nightly (fc2daaae6 2020-09-28)
```

### Error output

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `DefId(0:5 ~ playground[6bea]::bug[0]::{{opaque}}[0])`,
 right: `DefId(0:3 ~ playground[6bea]::bug[0])`', src/librustc_middle/hir/map/collector.rs:226:13
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
  Compiling playground v0.0.1 (/playground)
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:217
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:530
  12: rust_begin_unwind
             at src/libstd/panicking.rs:437
  13: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:391
  14: rustc_middle::hir::map::collector::NodeCollector::insert_entry
  15: <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_param
  16: rustc_hir::intravisit::walk_body
  17: rustc_hir::intravisit::walk_expr
  18: rustc_hir::intravisit::walk_block
  19: rustc_hir::intravisit::walk_expr
  20: rustc_hir::intravisit::walk_body
  21: rustc_hir::intravisit::walk_ty
  22: rustc_hir::intravisit::walk_assoc_type_binding
  23: rustc_hir::intravisit::walk_path
  24: rustc_hir::intravisit::Visitor::visit_param_bound
  25: rustc_hir::intravisit::walk_item
  26: <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_item
  27: rustc_hir::intravisit::walk_ty
  28: rustc_hir::intravisit::walk_fn
  29: <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_fn
  30: rustc_hir::intravisit::walk_item
  31: <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_item
  32: rustc_middle::hir::map::index_hir
  33: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  34: rustc_data_structures::stack::ensure_sufficient_stack
  35: rustc_query_system::query::plumbing::get_query_impl
  36: core::ops::function::FnOnce::call_once
  37: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  38: rustc_data_structures::stack::ensure_sufficient_stack
  39: rustc_query_system::query::plumbing::get_query_impl
  40: rustc_middle::hir::map::Map::expect_item
  41: rustc_middle::hir::map::Map::visit_item_likes_in_module
  42: rustc_passes::hir_id_validator::check_crate
  43: rustc_interface::passes::analysis
  44: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  45: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  46: rustc_query_system::query::plumbing::get_query_impl
  47: rustc_middle::ty::context::tls::enter_global
  48: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  49: rustc_span::with_source_map
  50: rustc_interface::interface::create_compiler_and_run
  51: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.46.0 (04488afe3 2020-08-24) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [index_hir] index HIR
#1 [hir_owner] HIR owner of `{{misc}}#0`
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `playground`.
```

</p>
</details>

### PR DESCRIPTION
**Pull request by lcnr**
fixes #77329

We now consistently make `impl Trait` a hir owner, requiring some special casing for synthetic generic params.

r? @eddyb

### CODE REVIEW

**Review by eddyb — APPROVED** [review]

File: src/test/ui/impl-trait/closure-in-impl-trait.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// run-pass
+#![allow(unused_must_use)]
+fn bug<T>() -> impl Iterator<Item = [(); { |x: u32| { x }; 4 }]> {
```
Comment (eddyb) [inline src/test/ui/impl-trait/closure-in-impl-trait.rs:3]: We're running into issues which suggest we also want examples where the `impl Trait` is in argument position.
(As that lowers to a generic parameter, which currently *isn't* a HIR owner)

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -508,6 +508,8 @@ impl Generics<'hir> {
 #[derive(HashStable_Generic)]
 pub enum SyntheticTyParamKind {
     ImplTrait,
+    // Created by the `#[rustc_synthetic]` attribute.
+    Rustc,
```
Comment (eddyb) [inline compiler/rustc_hir/src/hir.rs:512]: I'd call this `RustcSyntheticAttr` if you want to include `Rustc` in the name.

File: compiler/rustc_passes/src/hir_id_validator.rs [hunk 3]
```diff
@@ -163,4 +163,16 @@ impl<'a, 'hir> intravisit::Visitor<'hir> for HirIdValidator<'a, 'hir> {
...
         // different owner.
     }
+
+    fn visit_generic_param(&mut self, param: &'hir hir::GenericParam<'hir>) {
+        if let hir::GenericParamKind::Type {
+            synthetic: Some(hir::SyntheticTyParamKind::ImplTrait),
+            ..
+        } = param.kind
+        {
+            // Do nothing because bodging is fun.
```
Comment (eddyb) [inline compiler/rustc_passes/src/hir_id_validator.rs:173]: This needs a comment like `visit_impl_item_ref`, explaining that synthetic generics from `impl Trait` are HIR ID owners.

**Review by eddyb — APPROVED** [review]
r=me modulo comments

**Review by eddyb — APPROVED** [review]

**Comment by lcnr** [comment]
I can't produce this locally :/

**Comment by pietroalbini** [comment]
@bors r-

The PR failed but bors forgot about that during synchronize.

**Comment by lcnr** [comment]
> gzip: stdin: not in gzip format

okay, that looks spurious to me.

@bors retry

**Pull request merged**